desc/protoparse/ast/service.go

```diff
@@ -0,0 +1,178 @@
+package ast
+
+import "fmt"
+
+type ServiceNode struct {
+	compositeNode
+	Keyword    *KeywordNode
+	Name       *IdentNode
+	OpenBrace  *RuneNode
+	Options    []*OptionNode
+	RPCs       []*RPCNode
+	CloseBrace *RuneNode
+
+	AllDecls []ServiceElement
+}
+
+func (*ServiceNode) fileElement() {}
+
+func NewServiceNode(keyword *KeywordNode, name *IdentNode, open *RuneNode, decls []ServiceElement, close *RuneNode) *ServiceNode {
+	children := make([]Node, 4+len(decls))
+	children = append(children, keyword, name, open)
+	for _, decl := range decls {
+		children = append(children, decl)
+	}
+	children = append(children, close)
+
+	var opts []*OptionNode
+	var rpcs []*RPCNode
+	for _, decl := range decls {
+		switch decl := decl.(type) {
+		case *OptionNode:
+			opts = append(opts, decl)
+		case *RPCNode:
+			rpcs = append(rpcs, decl)
+		case *EmptyDeclNode:
+			// no-op
+		default:
+			panic(fmt.Sprintf("invalid ServiceElement type: %T", decl))
+		}
+	}
+
+	return &ServiceNode{
+		compositeNode: compositeNode{
+			children: children,
+		},
+		Keyword:    keyword,
+		Name:       name,
+		OpenBrace:  open,
+		Options:    opts,
+		RPCs:       rpcs,
+		CloseBrace: close,
+		AllDecls:   decls,
+	}
+}
+
+// ServiceElement is an interface implemented by all AST nodes that can
+// appear in the body of a service declaration.
+type ServiceElement interface {
+	Node
+	serviceElement()
+}
+
+var _ ServiceElement = (*OptionNode)(nil)
+var _ ServiceElement = (*RPCNode)(nil)
+var _ ServiceElement = (*EmptyDeclNode)(nil)
+
+type RPCNode struct {
+	compositeNode
+	Keyword    *KeywordNode
+	Name       *IdentNode
+	Input      *RPCTypeNode
+	Returns    *KeywordNode
+	Output     *RPCTypeNode
+	Semicolon  *RuneNode
+	OpenBrace  *RuneNode
+	Options    []*OptionNode
+	CloseBrace *RuneNode
+
+	AllDecls []RPCElement
+}
+
+func (n *RPCNode) serviceElement() {}
+
+func NewRPCNode(keyword *KeywordNode, name *IdentNode, input *RPCTypeNode, returns *KeywordNode, output *RPCTypeNode, semicolon, open *RuneNode, decls []RPCElement, close *RuneNode) *RPCNode {
+	if semicolon != nil {
+		if open != nil || len(decls) > 0 || close != nil {
+			panic("cannot specify both semicolon and RPC body nodes")
```

Comment:
The parser would not call it with inappropriate arguments. The grammar doesn't even allow it (to have both a semicolon _and_ a body for an rpc declaration).

Reply:
Basically, an error like this is caught during parsing, not during construction of a bad AST node. The panics are for misuse of the API, not because of bad input programs.

Reply:
BTW, if you think it would be more clear, I could separate the two cases into two different concrete node types -- like a `RPCNode` and `RPCBodyNode`, where one is used for the no-body form (that ends in a semi-colon) and the other is used for the body form (that has braces). Do you think that would be clearer?

Reply:
I left is as a single node type but split the factory function in two: `NewRPCNode` and `NewRPCNodeWithBody`